File: dotnet/src/Oppex.Integration.Sdk/IncidentClient.cs
using Microsoft.Extensions.Logging;
using Oppex.Integration.Sdk.Internal;

namespace Oppex.Integration.Sdk;

/// <summary>Posts incidents to Oppex.</summary>
/// <remarks>
/// <para>
/// The client is thread safe and intended to be shared. Create one per
/// application, reuse it, and dispose it during shutdown.
/// </para>
/// <code>
/// await using var client = new IncidentClient(new IncidentClientOptions
/// {
///     ApiKey = apiKey,
///     ServiceKey = serviceKey,
/// });
///
/// var response = await client.PostAsync(new IncidentRequest
/// {
///     Title = "Checkout latency breached the SLO",
///     Source = "checkout-api",
///     Severity = Severity.High,
/// });
/// </code>
/// </remarks>
public sealed class IncidentClient : IAsyncDisposable, IDisposable
{
    /// <summary>The Oppex incident endpoint every client posts to.</summary>
    public const string DefaultEndpoint = "https://api.oppex.ai/api/v1/incident/post";

    /// <summary>
    /// Redirects the whole delivery path at a loopback server. This exists so
    /// tests and the external smoke consumer can exercise the real transport; it
    /// is deliberately not an option on <see cref="IncidentClientOptions"/>,
    /// which would add a public knob that exists only to ease testing.
    /// </summary>
    private const string EndpointOverrideVariable = "OPPEX_TEST_ENDPOINT_URL";

    private readonly string? _serviceKey;
    private readonly ILogger _logger;
    private readonly HttpTransport _transport;
    private readonly AsyncDispatcher _dispatcher;
    private readonly TimeProvider _timeProvider;
    private readonly IReadOnlyList<TimeSpan> _retryDelays;
    private volatile bool _disposed;

    /// <summary>Creates a client.</summary>
    /// <exception cref="ArgumentException">The API key is missing or blank.</exception>
    public IncidentClient(IncidentClientOptions options)
        : this(options, TimeProvider.System, RetryPolicy.DefaultDelays)
    {
    }

    /// <summary>
    /// Test constructor. The retry schedule and clock are injectable only from
    /// inside this assembly, so the public surface never grows a retry knob.
    /// </summary>
    internal IncidentClient(IncidentClientOptions options, TimeProvider timeProvider,
        IReadOnlyList<TimeSpan> retryDelays)
    {
        ArgumentNullException.ThrowIfNull(options);
        if (string.IsNullOrWhiteSpace(options.ApiKey))
        {
            throw new ArgumentException("ApiKey must not be blank", nameof(options));
        }

        _serviceKey = string.IsNullOrWhiteSpace(options.ServiceKey) ? null : options.ServiceKey;
        _logger = options.Logger;
        _timeProvider = timeProvider;
        _retryDelays = retryDelays;
        _transport = new HttpTransport(options.ApiKey, ResolveEndpoint());
        _dispatcher = new AsyncDispatcher(_logger, timeProvider);
    }

    /// <summary>
    /// Delivers an incident and waits for the result, including any retry delays.
    /// The request's own service key overrides the client's.
    /// </summary>
    /// <exception cref="ArgumentException">
    /// The request is invalid, or neither the client nor the request carries a
    /// service key.
    /// </exception>
    /// <exception cref="ObjectDisposedException">The client was disposed.</exception>
    /// <exception cref="IncidentException">Delivery failed.</exception>
    public Task<IncidentResponse> PostAsync(IncidentRequest request, CancellationToken cancellationToken = default)
    {
        var normalized = Prepare(request);
        if (normalized.ServiceKey is null && _serviceKey is null)
        {
            throw new ArgumentException(
                "No ServiceKey is configured on the client or the request; supply one or use " +
                nameof(PostWithServiceRoutingAsync),
                nameof(request));
        }

        return DeliverAsync(normalized, _serviceKey, cancellationToken);
    }

    /// <summary>
    /// Delivers an incident without a service key so Oppex resolves the target
    /// service itself. The request must not carry its own service key. Otherwise
    /// identical to <see cref="PostAsync"/>.
    /// </summary>
    public Task<IncidentResponse> PostWithServiceRoutingAsync(
        IncidentRequest request, CancellationToken cancellationToken = default) =>
        DeliverAsync(PrepareForServiceRouting(request), null, cancellationToken);

    /// <summary>
    /// Queues a best-effort delivery and returns immediately.
    /// </summary>
    /// <remarks>
    /// Named <c>Enqueue</c> rather than following the Java SDK's <c>postAsync</c>:
    /// in .NET an <c>Async</c> suffix promises an awaitable, and this method
    /// deliberately is not one. Validation and the disposed check still run on the
    /// calling thread, so a misuse is thrown to the caller rather than lost in a
    /// worker. A delivery failure after queueing is logged at debug level.
    /// </remarks>
    public void Enqueue(IncidentRequest request)
    {
        var normalized = Prepare(request);
        if (normalized.ServiceKey is null && _serviceKey is null)
        {
            throw new ArgumentException(
                "No ServiceKey is configured on the client or the request; supply one or use " +
                nameof(EnqueueWithServiceRouting),
                nameof(request));
        }

        Submit(normalized, _serviceKey);
    }

    /// <summary>
    /// Queues a best-effort service-routed delivery. The request must not carry
    /// its own service key.
    /// </summary>
    public void EnqueueWithServiceRouting(IncidentRequest request) =>
        Submit(PrepareForServiceRouting(request), null);

    /// <summary>
    /// Drains queued work for up to ten seconds, then releases every owned
    /// resource. Idempotent.
    /// </summary>
    public async ValueTask DisposeAsync()
    {
        if (_disposed)
        {
            return;
        }

        _disposed = true;
        await _dispatcher.DisposeAsync().ConfigureAwait(false);
        _transport.Dispose();
    }

    /// <summary>
    /// Synchronous disposal, for callers that cannot await. Prefer
    /// <see cref="DisposeAsync"/>: this blocks the calling thread for the drain.
    /// </summary>
    public void Dispose() => DisposeAsync().AsTask().GetAwaiter().GetResult();

    private static Uri ResolveEndpoint()
    {
        var overrideUrl = Environment.GetEnvironmentVariable(EndpointOverrideVariable);
        return new Uri(string.IsNullOrEmpty(overrideUrl) ? DefaultEndpoint : overrideUrl);
    }

    private NormalizedRequest Prepare(IncidentRequest request)
    {
        ArgumentNullException.ThrowIfNull(request);
        ObjectDisposedException.ThrowIf(_disposed, this);
        return request.Normalize();
    }

    private NormalizedRequest PrepareForServiceRouting(IncidentRequest request)
    {
        var normalized = Prepare(request);
        return normalized.ServiceKey is null
            ? normalized
            : throw new ArgumentException(
                "Request must not carry a ServiceKey when service routing is used", nameof(request));
    }

    private void Submit(NormalizedRequest request, string? defaultServiceKey)
    {
        var accepted = _dispatcher.Submit(async cancellationToken =>
        {
            try
            {
                // No disposed check here: CloseAsync drains the dispatcher before
                // the transport is released, so a task queued before disposal
                // still has a live connection pool, and a task submitted after
                // disposal was already refused below.
                await DeliverAsync(request, defaultServiceKey, cancellationToken).ConfigureAwait(false);
            }
            catch (Exception failure) when (failure is IncidentException or OperationCanceledException)
            {
                Log.QueuedDeliveryFailed(_logger, failure);
            }
        });

        ObjectDisposedException.ThrowIf(!accepted, this);
    }

    /// <summary>
    /// Serializes, sends and retries a single incident. The request's own service
    /// key wins over <paramref name="defaultServiceKey"/>.
    /// </summary>
    private Task<IncidentResponse> DeliverAsync(
        NormalizedRequest request, string? defaultServiceKey, CancellationToken cancellationToken)
    {
        var payload = WireCodec.SerializeRequest(request, request.ServiceKey ?? defaultServiceKey);
        return RetryPolicy.ExecuteAsync(
            _retryDelays, token => _transport.SendAsync(payload, token), _timeProvider, cancellationToken);
    }
}

File: ruby/lib/oppex_sdk/incident_client.rb
# frozen_string_literal: true

require_relative "errors"
require_relative "incident_request"
require_relative "internal/async_dispatcher"
require_relative "internal/http_executor"
require_relative "internal/interrupt"
require_relative "internal/retry_executor"
require_relative "internal/wire_codec"
require_relative "logging"

module Oppex
  # Posts incidents to Oppex.
  #
  #   client = Oppex::IncidentClient.new(api_key: "api-key", service_key: "service-key")
  #   client.post(request)
  #   client.close
  #
  # The client is thread safe and intended to be shared. Create one per
  # application and close it during shutdown. It is also usable as a block:
  #
  #   Oppex::IncidentClient.open(api_key: "api-key") do |client|
  #     client.post_with_service_routing(request)
  #   end
  class IncidentClient
    # Creates a client, yields it, and closes it even if the block raises.
    #
    # @yieldparam client [IncidentClient]
    def self.open(...)
      client = new(...)
      begin
        yield client
      ensure
        client.close
      end
    end

    # @param api_key [String] sent in the +X-API-KEY+ header. Required.
    # @param service_key [String, nil] the default service for incidents that do
    #   not carry their own. When omitted, incidents must either supply one or be
    #   posted with {#post_with_service_routing}.
    # @param logger [#debug, #info, #warn, #error] receives this client's
    #   internal logging. Any object with those four methods works, including a
    #   standard +Logger+, +Rails.logger+, or your own.
    def initialize(api_key:, service_key: nil, logger: StderrLogger.new)
      @api_key = validate_api_key(api_key)
      @service_key = blank_to_nil(service_key, "service_key")
      @logger = logger
      @http_executor = Internal::HttpExecutor.new(@api_key)
      @interrupt = Internal::Interrupt.new
      @dispatcher = Internal::AsyncDispatcher.new(logger)
      @retry_delays = Internal::RetryExecutor::DEFAULT_DELAYS
      @closed = false
      @lifecycle = Mutex.new
    end

    # Posts on the calling thread, including any retry delays. The request's own
    # service key overrides the client's.
    #
    # @return [IncidentResponse]
    # @raise [ClientClosedError] after {#close}
    # @raise [ArgumentError] when neither the client nor the request carries a
    #   service key
    # @raise [IncidentError] when delivery fails
    def post(request)
      ensure_open!
      require_service_key!(request)
      deliver(request, @service_key)
    end

    # Posts without a service key so Oppex resolves the target service itself.
    # The request must not carry its own service key. Otherwise identical to
    # {#post}.
    #
    # @return [IncidentResponse]
    def post_with_service_routing(request)
      ensure_open!
      require_service_routable!(request)
      deliver(request, nil)
    end

    # Queues a best-effort delivery and returns immediately.
    #
    # The closed-client and service-key checks still run on the calling thread, so
    # a misuse is raised to the caller rather than lost in a worker. A delivery
    # failure after queueing is logged at debug level.
    #
    # @return [void]
    def post_async(request)
      ensure_open!
      require_service_key!(request)
      submit(request, @service_key)
    end

    # Queues a best-effort service-routed delivery. The request must not carry its
    # own service key.
    #
    # @return [void]
    def post_async_with_service_routing(request)
      ensure_open!
      require_service_routable!(request)
      submit(request, nil)
    end

    # @return [Boolean] whether {#close} has been called.
    def closed?
      @lifecycle.synchronize { @closed }
    end

    # Drains queued work for up to ten seconds, then releases every owned
    # resource. Idempotent, and safe to call from any thread.
    #
    # @return [void]
    def close
      @lifecycle.synchronize do
        return if @closed

        @closed = true
      end
      # Signalled before the drain so a worker sitting in an eight-second backoff
      # gives up now instead of consuming the whole drain budget.
      @interrupt.signal
      @dispatcher.close
      @http_executor.close
      nil
    end

    private

    def validate_api_key(api_key)
      raise TypeError, "api_key must be a String" unless api_key.is_a?(String)
      raise ArgumentError, "api_key must not be blank" if api_key.strip.empty?

      api_key
    end

    def blank_to_nil(value, field)
      return nil if value.nil?
      raise TypeError, "#{field} must be a String" unless value.is_a?(String)

      value.strip.empty? ? nil : value
    end

    def ensure_open!
      raise ClientClosedError if closed?
    end

    # Service routing is the only delivery mode available when no service key is
    # configured anywhere.
    def require_service_key!(request)
      return unless @service_key.nil? && request.service_key.nil?

      raise ArgumentError, "no service_key is configured on the client or the request; " \
                           "supply one or use post_with_service_routing"
    end

    def require_service_routable!(request)
      return if request.service_key.nil?

      raise ArgumentError, "request must not carry a service_key when service routing is used"
    end

    def submit(request, default_service_key)
      accepted = @dispatcher.submit do
        # No closed check here: close drains the dispatcher before releasing the
        # transport, so a task queued before close still has a live connection,
        # and a task submitted after close was already refused above.
        deliver(request, default_service_key)
      rescue IncidentError => e
        @logger.debug("oppex: asynchronous incident delivery failed: #{e.message}")
      end
      raise ClientClosedError unless accepted

      nil
    end

    # Serializes, sends and retries a single incident. The request's own service
    # key wins over +default_service_key+.
    def deliver(request, default_service_key)
      service_key = request.service_key || default_service_key
      payload = Internal::WireCodec.serialize_request(request, service_key)
      Internal::RetryExecutor.execute(@interrupt, delays: @retry_delays) do
        @http_executor.execute(payload)
      end
    end
  end
end
